headers/components/temp_doeclim_component.hpp

```diff
@@ -0,0 +1,150 @@
+ *  Adopted from C++ github implementation, cdice_doeclim (Garner et al., 2013).
+ */
+class TempDOECLIMComponent : public IModelComponent {
+    
+public:
+    TempDOECLIMComponent();
+    ~TempDOECLIMComponent();
+    
+    //! IModelComponent methods
+    virtual std::string getComponentName() const;
+    
+    virtual void init( Core* core );
+    
+    virtual unitval sendMessage( const std::string& message,
+                                const std::string& datum,
+                                const message_data info=message_data() ) throw ( h_exception );
+    
+    virtual void setData( const std::string& varName,
+                          const message_data& data ) throw ( h_exception );
+    
+    virtual void prepareToRun() throw ( h_exception );
+    
+    virtual void run( const double runToDate ) throw ( h_exception );
+    
+    virtual void shutDown();
+    
+    //! IVisitable methods
+    virtual void accept( AVisitor* visitor );
+    
+private:
+    virtual unitval getData( const std::string& varName,
+                            const double date ) throw ( h_exception );
+    void invert_1d_2x2_matrix( double * x, double * y);
+    void sum_1d_2x2_matrix( double * x, double * y, double * z);
+    
+    //Hard-coded DOECLIM parameters
+    int dt;                  // years per timestep (this is hard-coded into Hector)
+    int ns;                  // number of timesteps
+    double ak;               // slope in climate feedback - land-sea heat exchange linear relationship
+    double bk;               // offset in climate feedback - land-sea heat exchange linear relationship, W/m2/K
+    double csw;              // specific heat capacity of seawater W*yr/m3/K
+    double earth_area;       // m2
+    double kcon;             // conversion from cm2/s to m2/yr
+    double q2co;             // radiative forcing for atmospheric CO2 doubling, W/m2
+    double rlam;             // factor between land clim. sens. and sea surface clim. sens. T_L2x = rlam*T_S2x
+    double secs_per_Year;
+    double zbot;             // bottom depth of diffusive ocean, m
+    double bsi;              // warming factor for marine surface air over SST (due to retreating sea ice)
+    double cal;              // heat capacity of land-troposphere system, W*yr/m2/K
+    double cas;              // heat capacity of mixed layer-troposphere system, W*yr/m2/K
+    double flnd;             // fractional land area
+    double fso;              // ocean fractional area below 60m
+    
+    //Intermediate DOECLIM parameters
+    double B[4];
+    double C[4];
+
+    std::vector<double> KT0;
+    std::vector<double> KTA1;
+    std::vector<double> KTB1;
+    std::vector<double> KTA2;
+    std::vector<double> KTB2;
+    std::vector<double> KTA3;
+    std::vector<double> KTB3;
+    
+    std::vector<double> Ker;
+
+    double A[4];
+    double IB[4];
+    
+    //Dependent DOECLIM parameters
```

Comment:
Space after `//`. Dependent on what? What's the difference between `ocean_area` below and the same variable name above?

Reply:
@bpbond If these are just `consts` then we don't actually need them in the header?

Reply:
Mmm. Feeling fuzzy on my C++. I think actually we'd want `static const x = ...`. As for header or not, it depends; is it referred to in multiple component functions? If not, right, just move into the run method or wherever.

Reply:
`const` means that once it's initialized, it can't be changed.  `static` means that there is one copy for the entire class, instead of one for each object created.  A member can be either `const`, `static`, neither, or both, depending on what you are trying to accomplish.  For physical constants like the area of the earth or the specific heat of seawater, `static const` is the way to go.  For something that might be a settable parameter (`ak`, perhaps?), you would generally do that as a normal (i.e., not static) member.

A static member is initialized just like a global variable, irrespective of whether or not it happens to be const.  In fact, static members are basically just namespaced global variables, except that C++ does respect their `private` or `protected` visibility.

A member that is const, but not static _must_ be initialized in the constructor.  This is one of the rare cases in which the distinction between initialization and assignment matters.  In this case you have to use the `:` notation to make it an initialization instead of an assignment.  If you can't do that for some reason, such as because you have to construct the object before the initialization data is available, then you can't make the members const.  Your best bet in this case is to make them ordinary members and to add a comment to the effect that once they are set they should be treated as const.

A constant that is only used in one method could be moved into that method, as Ben suggests, but sometimes it's best to make them const or static const.  The reason why is that if someone wants to know, what are they using for the xyz parameter, it's a lot easier to find it if it's with all the other constants, rather than tucked away at the beginning of some particular method they might not know the name of.  Also, if you end up refactoring so that more than one method uses the constant, then you'll have to make the constant a member anyhow.  Starting it off that way future proofs your design a bit.